### PR TITLE
Create continuous-delivery-and-versioning-changes.md

### PR DESCRIPTION
**Pull request by matt-richardson**
# About

This is a post about our new continuous delivery approach and versioning changes.

# Pre-requisites

- [x] The draft is complete and this post is ready to be reviewed.
- [ ] I'm familiar with the [writing for Octopus TL;DR](https://style.octopus.com/writing-at-octopus-tldr).
- [x] The PR build passes validation, and if it doesn't, I've checked the [common validations errors](https://style.octopus.com/writing-at-octopus-tldr#common-validation-errors) and none of those apply.
- [ ] The images I've included follow the [rules for working with images](https://style.octopus.com/images#rules-for-screenshots).

## Publication date
I'm hoping to publish it asap.

### CODE REVIEW

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+---
+title: Continuous Delivery of Octopus
+description:
```
Comment (matt-richardson) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md:3]: Delivering Octopus Deploy faster and with higher quality

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+## Nightly builds and continuous deployment
+
+For a while now, we've been on a long-term missing to increase the flow of work and reduce time to feedback across all our teams. At the start of 2021, we were still facing some key constraints:
+
+* Our deployment pipelines frequently suffered bit rot
+* Our code changes often sat on branches that lived longer than we wanted
+
+As of this week, we've just delivered some key changes to help us ship faster and better quality releases.
+
+c ⏩ We are now practicing [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment) of Octopus Server, Tentacle and Octopus CLI to internal customer environments, and [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) to external customer environments. What does this mean you ask?
```
Comment (matt-richardson) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md:23]:
```suggestion
### ⏩ We are now practicing [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment) of Octopus Server, Tentacle and Octopus CLI to internal customer environments, and [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) to external customer environments. What does this mean you ask?
```

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+metaImage: to-be-added-by-marketing
+bannerImage: to-be-added-by-marketing
+tags:
+  - Product
+  - Engineering
+---
+
+## Nightly builds and continuous deployment
+
+For a while now, we've been on a long-term missing to increase the flow of work and reduce time to feedback across all our teams. At the start of 2021, we were still facing some key constraints:
```
Comment (michaelnoonan) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md]:
```suggestion
For a while now, we've been on a long-term mission to increase the flow of work and reduce time to feedback across all our teams. At the start of 2021, we were still facing some key constraints:
```

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+* Our deployment pipelines frequently suffered bit rot
```
Comment (michaelnoonan) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md]:
```suggestion
* Our deployment pipelines frequently suffered "bit rot"
```

Unsure if everyone knows what this means. :)

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+As of this week, we've just delivered some key changes to help us ship faster and better quality releases.
+
+### ⏩ We are now practicing [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment) of Octopus Server, Tentacle and Octopus CLI to internal customer environments, and [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) to external customer environments. What does this mean you ask?
+
+Well, instead of us having to make a deliberate decision to deploy a release, we now automatically create a release from every commit to a releasable branch that passes all the tests. Successful builds roll out to our internal environments and then after a suitable "bake time", onto Octopus Cloud, and then to the website. This means we are constantly "drinking our own champagne" in the pursuit of delivering a quality product.
+
+### 🌃 Nightly builds
+
+For our older LTS releases, we were finding that "bit rot" was causing our pipelines to fail as we were only exercising them when we needed to ship something. Now, we are triggering a build every night so that we know that when we need it, it's definitely working.
```
Comment (michaelnoonan) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md]:
```suggestion
For our older LTS releases, we were finding that "bit rot" was causing our pipelines to fail as we were only exercising them when we needed to ship a patch to an older release. Now, we are triggering a build every night to clean all the pipes, alerting us early when something goes wrong, which in turn makes the problem easier to diganose and cheaper to fix.
```

File: blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+Another interesting change in the mix here is around how we generate our release notes. We previously used GitHub milestones, and assigned them before we built a release. Now, as we're building a release for every commit, we are now using the Git revision graph to calculate what fixes went into what release. This means more accurate release notes.
+
+### 🚷 Better handling of valid/invalid upgrade path
+
+Now that we're using the Git revision graph to calculate these release notes, we also get some good knowledge about viable upgrade paths. Previously, it 2020.4.13 -> 2020.5.0 would appear to be a valid upgrade path, but would actually be going back in time (as 2020.5.0 was branched before 2020.4.13 was created). This occaisionally caused some bugs where the database had structural changes that were not expected. Now, we show a warning saying that this is not a viable upgrade path, meaning an entire class of bugs gets avoided.
+
+## Conclusion
+
+We're pretty excited about these changes - they'll help us focus less on the process and more on shipping good stuff.
```
Comment (michaelnoonan) [inline blog/2021-03/continuous-delivery-of-octopus/continuous-delivery-and-versioning-changes.md]:
```suggestion
We're pretty excited about these changes! They will help us focus more energy on improving Octopus, getting those improvements in Octopus Cloud customer's hands sooner, and give our customers a better experience when upgrading their self-hosted installation.
```

**Review by michaelnoonan — APPROVED** [review]

**Comment by wordlee** [comment]
It looks like we had a midair crossover. I'll pull Mike's comments into the updated file.

File: blog/2021-03/continuous-delivery-of-octopus/index.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
+---
+title: Continuous Delivery of Octopus
+description: Delivering Octopus Deploy faster and with higher quality
```
Comment (uglybugger) [inline blog/2021-03/continuous-delivery-of-octopus/index.md:3]: Nit: adjective versus adverb. "Faster" is an adjective and describes a noun; "quickly" is an adverb and describes a verb.
``
```suggestion
description: Delivering Octopus Deploy more quickly and with higher quality
```

File: blog/2021-03/continuous-delivery-of-octopus/index.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+## Nightly builds and continuous deployment
+
+We've been on a long-term mission for sometime to increase the flow of work and reduce time to feedback across all our teams. At the start of 2021, we still faced some key constraints:
+
+* Our deployment pipelines frequently suffered "bit rot".
+* Our code changes often sat on branches that lived longer than we wanted.
+
+As of this week, we've delivered some key changes to help us ship faster and better quality releases.
+
+### ⏩ We are now practicing [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment) of Octopus Server, Tentacle and Octopus CLI to internal customer environments, and [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) to external customer environments. What does this mean you ask?
```
Comment (uglybugger) [inline blog/2021-03/continuous-delivery-of-octopus/index.md:25]: Grammatical nit: quoting the question.

```suggestion
### ⏩ We are now practicing [Continuous Deployment](https://en.wikipedia.org/wiki/Continuous_deployment) of Octopus Server, Tentacle and Octopus CLI to internal customer environments, and [Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) to external customer environments. "What does this mean?" you ask.
```

File: blog/2021-03/continuous-delivery-of-octopus/index.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+Well, instead of us having to make a deliberate decision to deploy a release, we now automatically create a release from every commit to a releasable branch that passes all the tests. Successful builds roll out to our internal environments and then after a suitable "bake time", onto Octopus Cloud, and then to the website. This means we are constantly "drinking our own champagne" in the pursuit of delivering a quality product.
+
+You can read more about how we see the differences between [continuous delivery and continuous deployment](blog/2021-03/ten-pillars-of-pragmatic-deployments#continuous-integration-continuous-delivery-and-continuous-deployment.md).
+
+### 🌃 Nightly builds
+
+For our older LTS releases, we found that "bit rot" was causing our pipelines to fail as we only exercised them when we needed to ship a patch to an older release. Now, we trigger a build every night to clean all the pipes, alerting us early when something goes wrong, which in turn makes the problem easier to diagnose and cheaper to fix.
```
Comment (uglybugger) [inline blog/2021-03/continuous-delivery-of-octopus/index.md:33]:
```suggestion
For our older LTS releases, we found that "bit rot" was causing our pipelines to fail as we only exercised them when we needed to ship a patch to an older release. Now, in addition to our commit-triggered builds, we trigger a build every night for every potentially-releasable branch to clean all the pipes, alerting us early when something goes wrong, which in turn makes the problem easier to diagnose and cheaper to fix.
```

File: blog/2021-03/continuous-delivery-of-octopus/index.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+### 🔢 major.minor.build versioning
+
+Nightly builds did surface an interesting challenge for us though. With our prior numbering scheme, rebuilding every night would result in the same version number. Many downstream systems (e.g. nuget.org) don't like different packages with the same version number, which meant we needed a new plan. We now use a major.minor.build numbering strategy. This means the build number are much larger than we're all used to, but effectively, it's just a number. As this is now a build number, there will be gaps between version numbers, but you'll almost always want to grab the latest build.
```
Comment (uglybugger) [inline blog/2021-03/continuous-delivery-of-octopus/index.md:37]:
```suggestion
Nightly builds did surface an interesting challenge for us, though. With our prior numbering scheme (`major.minor.patch`), rebuilding every night would result in multiple builds per Git commit, and our build numbers would only change if a commit had been made. This meant that we ended up with multiple builds with the same version number. Many downstream systems (e.g. nuget.org, registry.npmjs.org) don't like different packages with the same version number, which meant we needed a new plan.

We now use a `major.minor.build` numbering strategy. This means the patch numbers are much larger than we're all used to but, effectively, it's just a number. As this is now a build number, there will be gaps between version numbers, but you'll almost always want to grab the latest build. 
```

File: blog/2021-03/continuous-delivery-of-octopus/index.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+Another interesting change in the mix here is around how we generate our release notes. We previously used GitHub milestones and assigned them before we built a release. Now, as we're building a release for every commit, we use the Git revision graph to calculate what fixes went into what release. This means more accurate release notes.
+
+### 🚷 Better handling of valid/invalid upgrade path
+
+Now that we use the Git revision graph to calculate these release notes, we also get some good knowledge about viable upgrade paths. Previously, 2020.4.13 -> 2020.5.0 would appear to be a valid upgrade path, but would actually go back in time, as 2020.5.0 was branched before 2020.4.13 was created. This occasionally caused some bugs where the database had structural changes that weren't expected. Now, we show a warning that this isn't a viable upgrade path, meaning an entire class of bugs is avoided.
+
+## Conclusion
+
+We're pretty excited about these changes! They will help us focus more energy on improving Octopus, getting those improvements in Octopus Cloud customer's hands sooner, and give our customers a better experience when upgrading their self-hosted installation.
```
Comment (uglybugger) [inline blog/2021-03/continuous-delivery-of-octopus/index.md:51]:
```suggestion
We're pretty excited about these changes! They will help us focus more energy on improving Octopus, getting those improvements in Octopus Cloud customers' hands sooner, and give our customers a better experience when upgrading their self-hosted installations.
```

File: blog/2021-03/continuous-delivery-of-octopus/index.md (hunk 7 above)
Comment (wordlee) [reply blog/2021-03/continuous-delivery-of-octopus/index.md:3]: How does this sound:
How we're delivering higher quality releases of Octopus Deploy in less time.

File: blog/2021-03/continuous-delivery-of-octopus/index.md (hunk 8 above)
Comment (wordlee) [reply blog/2021-03/continuous-delivery-of-octopus/index.md:25]: This feels a little bit awkward. How about changing that to:

What does this mean for you?

File: blog/2021-03/continuous-delivery-of-octopus/index.md (hunk 7 above)
Comment (matt-richardson) [reply blog/2021-03/continuous-delivery-of-octopus/index.md:3]: I like yours there, @wordlee 👍

File: blog/2021-03/continuous-delivery-of-octopus/index.md (hunk 8 above)
Comment (matt-richardson) [reply blog/2021-03/continuous-delivery-of-octopus/index.md:25]: Not fussed - either way 👍

**Comment by wordlee** [comment]
This branch is missing a file we link to from the body of the text. The validation issues should be resolved by merging to master.

**Pull request merged**